### PR TITLE
Sort out ARM load/store instruction size issues

### PR DESCRIPTION
**Pull request by mikedn**
This fixes a number of inconsistencies in ARM 32/64's handling of loads (and to a lesser extent of stores as well).

For a start, ARM32 and 64 have a different approach to load instruction size (`emitAttr`). The ARM32 emitter wants `EA_1BYTE`/`EA_2BYTE` for byte/halfword loads. On the other hand, the ARM64 emitter expects `EA_4BYTE` or `EA_8BYTE`, to it the instruction size indicates the size of the destination register and not the size of the memory location (that is implied by the instruction anyway). It makes more sense for ARM32 to follow the ARM64 approach and require `EA_4BYTE` for small loads, at least because it avoids the need to write different code for ARM32 and ARM64. Besides, ARM32 doesn't really have `EA_1BYTE`/`EA_2BYTE` instructions, they're all 4 byte instructions.

The only reason why the ARM32 emitter cares about the load instruction size is immediate operand scaling. Of course, the scale factor can be deduce from the instruction itself and the size becomes is irrelevant. Asserting that it's `EA_4BYTE` just to enforce consistency.

Then ARM64 itself handles load size in an unusual manner. Instead of simply using `emitActualTypeSize` when calling `emitInsLoadStoreOp`, it uses `emitTypeSize` and then `emitInsLoadStoreOp` fixes the size using `emitInsAdjustLoadStoreAttr`. And that not only promotes `EA_1BYTE`/`EA_2BYTE` to `EA_4BYTE`, it also promotes `EA_4BYTE` to `EA_8BYTE`. Long story short - calling `emitInsLoadStoreOp(INS_ldr, EA_4BYTE)` results in a 8 byte `ldr` being unexpectedly emitted!

And the reason for `emitInsAdjustLoadStoreAttr` weird behavior is another ARM64 inconsistency - its `ins_Load` returns `INS_ldrsw` for `TYP_INT`. And `INS_ldrsw` requires `EA_8BYTE` so `emitInsAdjustLoadStoreAttr` addresses that need. But the use of `INS_ldrsw` is actually bogus, it should be just `INS_ldr`:
* the JIT's type model doesn't require that a `TYP_INT` load sign extends the 32 bit value to 64 bit
* the use of `TYP_INT` does not imply that the loaded value is actually signed
* XARCH emits `mov`, not `movsxd` for `TYP_INT` loads

So change `ins_Load` to return `INS_ldr`, not `INS_ldrsw`, and get rid of `emitInsAdjustLoadStoreAttr`.

Along the way also remove some `ins_Load(TYP_INT)` and similar that just make the code harder to read.

And then fix `genCodeForIndexAddr` which tries to sign extend the array length unnecessarily and then always emits a 64 bit cmp, disregarding the index type.

And fix the volatile load/store logic too, which got caught into this `TYP_INT` load confusion and failed to actually emit `INS_ldar`, despite trying to do so.

Oh well, I suppose it's kind of confusing. I'm just trying to finish my cast operand containment thing and it seems that I keep running into all kind of weird things...

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked crossgen_comparison Build and Test
@dotnet-bot test Ubuntu arm Cross Release crossgen_comparison Build and Test

**Comment by mikedn** [comment]
No FX diffs on ARM32. On ARM64 there are many diffs due to the `ldrsw` -> `ldr` change
```diff
 G_M55349_IG12:
-            ldrsw   x0, [fp,#24]	// [V03 loc1]
+            ldr     w0, [fp,#24]	// [V03 loc1]
             tst     w0, #255
             beq     G_M55349_IG13
```
a few due to byte/halfword instructions becoming EA_4BYTE instead of EA_8BYTE
```diff
             ldr     x1, [fp,#104]	// [V04 loc0]
-            ldrsh   x1, [x1]
+            ldrsh   w1, [x1]
             strh    w1, [x0,#8]
             b       G_M14724_IG34
```
and some code size improvements due to the use of ldar in many more places:
```diff
             mov     x0, x21
-            ldrsw   x22, [x0]
-            dmb     oshld
-            ldrsw   x0, [x19,#24]
+            ldar    w22, [x0]
+            ldr     w0, [x19,#24]
             and     w23, w22, w0
             ldr     x0, [x19,#8]
```
These account for ~1.5 kbytes of diff
```
Total bytes of diff: -1504 (0.00% of base)
    diff is an improvement.
Top file improvements by size (bytes):
        -872 : System.Private.CoreLib.dasm (-0.01% of base)
        -496 : Microsoft.CodeAnalysis.CSharp.dasm (0.00% of base)
         -80 : System.Net.Http.dasm (-0.01% of base)
         -32 : Microsoft.CodeAnalysis.VisualBasic.dasm (0.00% of base)
         -16 : System.Linq.Parallel.dasm (0.00% of base)
6 total files with size differences (6 improved, 0 regressed), 124 unchanged.
Top method improvements by size (bytes):
        -256 : System.Private.CoreLib.dasm - ConcurrentQueue`1:get_Count():int:this (4 methods)
         -48 : Microsoft.CodeAnalysis.CSharp.dasm - SymbolCompletionState:SpinWaitComplete(int,struct):this (4 methods)
         -48 : System.Net.Http.dasm - ConcurrentQueueSegment`1:TryDequeue(byref):bool:this (4 methods)
         -48 : System.Private.CoreLib.dasm - WorkStealingQueue:LocalPush(ref):this (2 methods)
         -32 : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:ForceComplete(ref,struct):this (4 methods)
73 total methods with size differences (73 improved, 0 regressed), 113019 unchanged.
108 files had text diffs but not size diffs.
System.Private.Xml.dasm had 85832 diffs
Microsoft.CodeAnalysis.dasm had 41384 diffs
Microsoft.Diagnostics.Tracing.TraceEvent.dasm had 38596 diffs
System.Private.DataContractSerialization.dasm had 18556 diffs
System.Text.RegularExpressions.dasm had 17844 diffs
```

**Comment by mikedn** [comment]
@BruceForstall Any idea what exactly failed in the arm build? Looks like all test passed but xunit exited with code 1?!

**Comment by echesakovMSFT** [comment]
@mikedn It failed to load a  Xunit test wrapper (https://github.com/dotnet/coreclr/issues/20392)
```
13:58:25   System.IO.FileLoadException: Could not load file or assembly 'baseservices.varargs.XUnitWrapper, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'. An operation is not legal in the current state. (Exception from HRESULT: 0x80131509 (COR_E_INVALIDOPERATION))
```

**Comment by mikedn** [comment]
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test

Assembly load failure again...

**Comment by mikedn** [comment]
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test

```
Groovy script failed:
java.lang.NullPointerException: Cannot invoke method child() on null object
	at org.codehaus.groovy.runtime.NullObject.invokeMethod(NullObject.java:91)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:48)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.NullCallSite.call(NullCallSite.java:35)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:155)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:157)
	at org.kohsuke.groovy.sandbox.impl.Checker$checkedCall$0.callStatic(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:56)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:194)
	at Script1.run(Script1.groovy:10)
```
Ain't that groovy?

**Comment by mikedn** [comment]
ARM64 FX minopts diff:
```
Total bytes of diff: -384 (0.00% of base)
    diff is an improvement.
Top file improvements by size (bytes):
        -224 : System.Net.NetworkInformation.dasm (-0.13% of base)
        -160 : System.Private.CoreLib.dasm (0.00% of base)
2 total files with size differences (2 improved, 0 regressed), 128 unchanged.
Top method improvements by size (bytes):
         -16 : System.Private.CoreLib.dasm - Volatile:Read(byref):int (4 methods)
         -16 : System.Private.CoreLib.dasm - SpinLock:ContinueTryEnter(int,byref):this (2 methods)
         -16 : System.Private.CoreLib.dasm - SpinLock:ContinueTryEnterWithThreadTracking(int,int,byref):this (2 methods)
         -16 : System.Private.CoreLib.dasm - SpinLock:ExitSlowPath(bool):this (2 methods)
         -16 : System.Private.CoreLib.dasm - SpinLock:get_IsHeld():bool:this (2 methods)
43 total methods with size differences (43 improved, 0 regressed), 113273 unchanged.
117 files had text diffs but not size diffs.
System.Private.Xml.dasm had 186632 diffs
Microsoft.CodeAnalysis.VisualBasic.dasm had 174484 diffs
Microsoft.CodeAnalysis.CSharp.dasm had 142040 diffs
Microsoft.CodeAnalysis.dasm had 72784 diffs
Microsoft.Diagnostics.Tracing.TraceEvent.dasm had 71612 diffs
```
In minopts there are additional diffs due to the `genCodeForIndexAddr` change:
```diff
-            ldrsw   x2, [x0,#8]
-            cmp     x1, x2
+            ldr     w2, [x0,#8]
+            cmp     w1, w2
```

**Comment by BruceForstall** [comment]
@dotnet-bot test this please

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

**Comment by alpencolt** [comment]
cc @dotnet/arm32-contrib @kbaladurin @gbalykov

File: src/jit/emitarm.cpp [hunk 1]
```diff
@@ -4928,31 +4917,44 @@ inline unsigned insEncodeImmT2_Mov(int imm)
...
+//    fmt - the instruction format
+//    imm - the immediate value to unscale
+//
+// Return Value:
+//    The unscaled immediate value
+//
+// Notes:
+//    Only IF_T1_C encoded instructions are supported.
+//
+/*static*/ int emitter::insUnscaleImm(instruction ins, insFormat fmt, int imm)
```
Comment (briansull) [inline src/jit/emitarm.cpp]: Seems a bit odd the have a parameter that is always required to be IF_T1_C

Do you believe that we will extend support to include more that one instruction format here?

Comment (mikedn) [reply src/jit/emitarm.cpp]: Right, it's unlikely to add more instructions formats here.

AFAIR this is just a side effect of the way I refactored the code - first I added `ins` + `fmt` and removed `size` parameter, at that point it was still supporting more formats. Then I realized that all formats but `IF_T1_C` had trivial scaling needs, similar to that of many other formats that did some kind of scaling without using `insUnscaleImm`. So I removed everything but `IF_T1_C` and added an assert to be sure the function doesn't get misused by accident.

I suppose I could go one step further - rename it to `insUnscaleImm_IF_T1_C` and remove the format parameter?

**Comment by mikedn** [comment]
`genCodeForIndexAddr` needs an extra fix, like the XARCH version it fails to widen the index:
```C#
    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Test(Foo[] a, long i)
    {
        return a[(int)checked(i + 1)].x;
    }
```
generates
```asm
        F9400FA0          ldr     x0, [fp,#24]
        F9400BA1          ldr     x1, [fp,#16]
        52800022          mov     w2, #1
        93407C42          sxtw    x2, x2
; generates a 64 bit value in x1
        AB020021          adds    x1, x1, x2
        54000146          bvs     G_M55886_IG04
        B9400802          ldr     w2, [x0,#8]
        6B02003F          cmp     w1, w2
        54000102          bhs     G_M55886_IG05
        52800622          mov     w2, #49
; the 64 bit value in x1 is used as is, the cast to int was dropped
        9B020020          madd    x0, x1, x2, x0
        91004000          add     x0, x0, #16
        B9400000          ldr     w0, [x0]
```

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib Take for 3.0 or hold off?

**Comment by CarolEidt** [comment]
I would be happy to take this for 3.0, as the code issue has been identified as an issue, and this is also some good code cleanup. I'd like to get @briansull to weigh in on it as well (i.e. is there more risk that I might think?) and the conflicts will need to be resolved.

**Comment by briansull** [comment]
I am fine with taking this change for 3.0
It does improve things quite a bit when selecting the proper load/store instructions for the ARM architectures.

File: src/jit/emitarm.cpp (hunk 1 above)
Comment (briansull) [reply src/jit/emitarm.cpp]: I don't think that it needs to be renamed, the extra parameter can just be removed.
We still have this assert which prevents unexpected callers from using this method:

```
       default:
            assert(!"Invalid value in size");	            assert(!"Invalid IF_T1_C instruction");
```

**Comment by mikedn** [comment]
What should be done about the index widening issue in `genCodeForIndexAddr` that I mentioned in a previous post? I've been looking at fixing that but I don't see a way to fix it without an extra temp register which means that this change will spread to lsraarm64.cpp as well. Perhaps that would be best done in a separate PR?

**Comment by CarolEidt** [comment]
> What should be done about the index widening issue in genCodeForIndexAddr that I mentioned in a previous post? I've been looking at fixing that but I don't see a way to fix it without an extra temp register which means that this change will spread to lsraarm64.cpp as well. Perhaps that would be best done in a separate PR?

I missed that this was a bug, rather than a CQ issue. I think it's reasonable to do that in a separate PR. It's a pretty corner case, and it seems that although the bug was mentioned in #22528, it doesn't appear that a test case was ever constructed to demonstrate the issue.

**Comment by briansull** [comment]
I thought that it was already fixed with:

**Fix genCodeForIndexAddr** #22528

**Comment by CarolEidt** [comment]
> I thought that it was already fixed with: Fix genCodeForIndexAddr #22528

That fixed it for x64, not for arm64.

**Comment by mikedn** [comment]
Yes, #22528 was only for x64. At the time this PR was already opened and I was hoping to make the ARM64 fix here. But then it turned out that the ARM64 fix is more complex and then this PR didn't get much attention so I kept postponing doing it.

> It's a pretty corner case, and it seems that although the bug was mentioned in #22528, it doesn't appear that a test case was ever constructed to demonstrate the issue.

Yes, it seems difficult to construct a reasonable example that exhibits observable incorrect behavior. The little example above does show that incorrect code is being generated but it's already very contrived. In particular, I haven't been able to find any example that would result in an out of range access due to an incorrect range check.

I'll fix conflicts later today.

File: src/jit/emitarm.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/emitarm.cpp]: Right, parameter removed.

**Comment by mikedn** [comment]
One thing that it's worth pointing out is that `ins_Load` & co. are inherently error prone. They return an instruction for a given type but they do not also return an `emitAttr` for that type and the caller has to ensure that the proper `emitAttr` is used.

Future cleanup of the emitter might consider packing the instruction and the attr in a struct (containing an unsigned bitfield for example) and making `ins_Load` & co. return such a struct. Emitter functions will have one parameter less and the chance to use the wrong attr will be smaller.

**Comment by CarolEidt** [comment]
Merging, since the two failed legs are x64, and all the changes are arm/arm64 specific.
